### DISCUSSION TITLE
Flickering of lines while rendering

### DISCUSSION THREAD

**Discussion by ComradeMashkov** [pelicanmapping/rocky#62] [Q&A / Support] [answered]
I am experiencing an issue in my program where lines appear to flicker or jitter when rendered. I have attached a gif that demonstrates the problem. Could you please help investigate what might be causing this?

[gif](https://pouch.jumpshare.com/preview/OpUsc9v6i7g7M4zgmUQDLH3ewzoZBCXAMQ2Anywk2-1Lo5kE-iwwIbr4a1BJ-T6N6M5SCjY79pIJ8sGkOqsJgsl4_wnASYOV_NuT5Ihr3_o)

**Comment by gwaldron** [comment]
Your link gives a 404. Can you try again?

**Reply by ComradeMashkov** [reply]
Sorry, my fault.

[gif](https://imgur.com/gallery/lines-flickering-AkKPMbI)

**Comment by gwaldron** [comment]
Thanks. This is happening due to floating-point precision loss when using ECEF coordinates in the geometry.
We are working on another API call Line (et al) that lets you push the verts with geopoints -- this will eliminate the issue.
Stay tuned...this will be fixed soon.

**Comment by gwaldron** [comment]
I pushed a commit that should resolve this issue. The API has changed though, so please refer to `Demo_Line.h` to see how it's done.
You need to call `setReferencePoint` and then use the returned `SRSOperation` to submit your points.
Commit 9fdb53e

**Reply by ComradeMashkov** [reply]
Thanks for your fast reply! However, it didn't resolve my problem. My rendered lines are still flickering during camera rotation & movement. I don't get what is the root of this problem... Maybe I'm doing something wrong?

```cpp
    // First, create the line entity
    entt::entity lineEntity = entt::null;

    lineEntity = app.entities.create();

    rocky::Line& line = app.entities.emplace<rocky::Line>(lineEntity);

    // Set a reference point. This should be near your geometry, and will
    // act as an anchor point for localizing geometry.
    // Use the returned SRSOperation to transform points for use in your line.
    rocky::GeoPoint refPoint(rocky::SRS::WGS84, 50.103129, 26.327247);
    auto lineXform = line.setReferencePoint(refPoint);
    std::vector<glm::dvec3> points;

    const std::string path_to_coords = native_dir + "data\\coords\\global_trajectories2.csv";
    const std::vector<double> latitude = utils::GetPoints(path_to_coords, "lat", id, 10);
    const std::vector<double> longitude = utils::GetPoints(path_to_coords, "lon", id, 10);
    const std::vector<double> altitude = utils::GetPoints(path_to_coords, "z", id, 10);

    for (int i = 0; i < latitude.size(); ++i)
    {
        glm::dvec3 point; 
        if (lineXform.transform(glm::dvec3(longitude[i], latitude[i], altitude[i]), point))
        {
            points.emplace_back(point);
        }
    }
    
    line.push(points.begin(), points.end());

    line.style = rocky::LineStyle();
    line.style->color = color;
    line.style->width = 2.0f;
    line.style->stipple_pattern = 0xffff;
    line.style->stipple_factor = 1;
    line.write_depth = true;
```

**Comment by gwaldron** [comment]
Hmm, and you are sure the reference point is reasonably close to your data points?
Maybe you can upload a small data sample for us to test?

**Reply by ComradeMashkov** [reply]
Yes, I've double-checked it, but I can be wrong ofc. Here is the data for plotting points for one line.
[test.csv](https://github.com/user-attachments/files/17511765/test.csv)

**Comment by gwaldron** [comment]
Thanks, that helps. Try a new pull - should be better.

**Reply by ComradeMashkov** [reply]
Thank you! That helped a lot, it solved an issue. Works great!

[result](https://imgur.com/gallery/flickering-test-ThrPgRL)

**Comment by gwaldron** [comment, marked as answer]
Update:
I had to change the API again, so there's a new answer to this question. Such is life with a pre-1.0 SDK :)
We made these changes while refining the ECS architecture.

The old approach (above) included calling `setReferencePoint` and using the returned `SRSOperation`.

The new approach is similar but simpler:
```c++
line.referencePoint = GeoPoint(...);
```
Then, just add a vector of points, expressed in the SRS of the reference point (e.g., lon, lat, alt).

**Reply by ComradeMashkov** [reply]
Thanks! Working as good as always.
